File: htf_impact_analysis.jl
import SimpleFeatures as SF
using DataFrames
import ArchGDAL as AG
using ProgressMeter
using GeoInterface

#------------------------- Measuring and summarizing impact of HTF on roadways -------------------
# Load road segments. We will use the 'road_segments' for every calculation
road_segments = SF.st_read("/Users/adam/Downloads/osm_edges.gpkg")

# # Load buffered roadways
# buffered_roads = SF.st_read("data/nc_road_lines_w_nodes_proj.gpkg")

# htf_polygons = SF.st_read("/Users/adam/Downloads/htf_on_rds.fgb")

# Define our 'impact' function. Here's how it works:
#   STEPS
#   ------
#   1) Based on buffered roadways and HTF, we are going 
#   to find the roadway segments that may be impacted by HTF (based on broader 'id').
#
#   2) We will filter down more to find the specific road segments that intersect HTF, and then buffer them accordingly
#
#   3) We will use the road centerline of each impacted road segment, to create perpindicular lines every 1 meter that 
#   exactly span the roadway.
#
#   4) HTF polygons contain the road id (:id) and the confidence class (low or high confidence). We want to estimate 
#   the roadway width coverage of each confidence class of HTF. We do this by taking the length of each perpindicular line 
#   across the road, erasing the high confidence HTF polygons from the line, finding the new length, then doing the same
#   thing with the low confidence HTF polygons. Each perpindicular roadway line now has a high confidence HTF roadway width
#   coverage and a low confidence HTF roadway coverage that are in % of roadway width. These will be reported as a range of
#   minimum roadway coverage (high confidence) and maximum roadway coverage (high confidence + low confidence)

function classify_road_flooding(;road_segments::SF.SimpleFeature, htf_polygons::SF.SimpleFeature)

    roads_affected = Set(htf_polygons.df[:,id_column])
    
    filtered_road_segments = @rsubset(road_segments, :id in roads_affected)
    replace!(filtered_road_segments.df.lanes, missing => "2")
    @rtransform!(filtered_road_segments, :lanes = tryparse(Int, :lanes))
    replace!(filtered_road_segments.df.lanes, nothing => 2)

    @transform!(filtered_road_segments, :buffer_distance = 2.5 * :lanes)

    filtered_road_segments_df = SF.sf_to_df(filtered_road_segments)
    # filtered_road_segments[!, :geom] = SF.from_sfgeom(filtered_road_segments.df.geom, to = "gdal")

    htf_polygons_df = SF.sf_to_df(htf_polygons);

    high_confidence_htf_polygons_df = SF.sf_to_df(SF.st_combine(@rsubset(htf_polygons[:,["DN","geom"]], :DN === 1)))
    low_confidence_htf_polygons_df = SF.sf_to_df(@rsubset(htf_polygons, :DN === 0))
    
    road_segments_impacted = DataFrame()

    @showprogress for row in eachrow(htf_polygons_df)
        for i in 1:nrow(filtered_buffered_roads)
            if AG.intersects(row.geom, AG.buffer(filtered_road_segments.geom[i], filtered_road_segments.buffer_distance[i])) == true
                append!(road_segments_impacted, DataFrame(filtered_road_segments[i,:]))
            end
        end
    end

    impact_classified_roads = DataFrame()

    @showprogress for row in eachrow(road_segments_impacted)
        road = row.geom
        buffered_road = AG.buffer(row.geom, row.buffer_distance)

        road_coords = GeoInterface.coordinates(road)

        rise_run = (road_coords[2] - road_coords[1])
        perp_angle = (-1 * rise_run[1])/rise_run[2]
    
        line_clone = AG.clone(road)
        pt_coords = GeoInterface.coordinates(AG.segmentize!(line_clone, 1))
    
        line_list = []
        hc_erased_line_list = []
        lc_erased_line_list = []

        for i in pt_coords
            new_line = AG.createlinestring()
            AG.addpoint!(new_line, i[1] + (row.buffer_distance[1]), i[2] + ((row.buffer_distance[1]) * perp_angle))
            AG.addpoint!(new_line, i[1] - (row.buffer_distance[1]), i[2] - ((row.buffer_distance[1]) * perp_angle))

            new_line_int = AG.intersection(new_line, buffered_road)
            push!(line_list, new_line_int)

            if nrow(high_confidence_htf_polygons_df) > 0
                hc_erased_line = AG.difference(new_line_int, high_confidence_htf_polygons_df.geom[1])
                push!(hc_erased_line_list, hc_erased_line)
            else
                push!(hc_erased_line_list, new_line_int)
            end

            if nrow(low_confidence_htf_polygons_df) > 0
                lc_erased_line = AG.difference(new_line_int, low_confidence_htf_polygons_df.geom[1])
                push!(lc_erased_line_list, lc_erased_line)
            else
                push!(lc_erased_line_list, new_line_int)
            end
        end    
    
        new_df = repeat(select(DataFrame(row), Not(:geom)), length(line_list))
        new_df.length = AG.geomlength.(line_list)
        new_df.hc_length = AG.geomlength.(hc_erased_line_list)
        new_df.lc_length = AG.geomlength.(lc_erased_line_list)

        new_df.hc_percent_covered = 100 .- ((new_df.hc_length./new_df.length)*100)
        new_df.lc_percent_covered = 100 .- ((new_df.lc_length./new_df.length)*100)
        
        new_df.min_percent_covered = new_df.hc_percent_covered
        new_df.max_percent_covered = new_df.hc_percent_covered + new_df.lc_percent_covered

        n_min_75_100 = sum(new_df.min_percent_covered .>= 75)
        n_min_50_75 = sum(new_df.min_percent_covered .>= 50 .&& new_df.min_percent_covered .< 75)
        n_min_25_50 = sum(new_df.min_percent_covered .>= 25 .&& new_df.min_percent_covered .< 50)
        n_min_0_25 = sum(new_df.min_percent_covered .> 0 .&& new_df.min_percent_covered .< 25)
        n_min_0 = sum(new_df.min_percent_covered .== 0.0)

        n_max_75_100 = sum(new_df.max_percent_covered .>= 75)
        n_max_50_75 = sum(new_df.max_percent_covered .>= 50 .&& new_df.max_percent_covered .< 75)
        n_max_25_50 = sum(new_df.max_percent_covered .>= 25 .&& new_df.max_percent_covered .< 50)
        n_max_0_25 = sum(new_df.max_percent_covered .> 0 .&& new_df.max_percent_covered .< 25)
        n_max_0 = sum(new_df.max_percent_covered .== 0.0)

        # Impact class: 3 = 75-100% coverage of HTF on road, 2 = 50 - 75%, 1 = 25 - 50%, 0 = 0 - 25%
        min_impact_class = [ifelse(n_min_75_100 > 2, 3, 
        ifelse(n_min_50_75 > 2, 2,
        ifelse(n_min_25_50 > 2, 1, 0)))]

        max_impact_class = [ifelse(n_max_75_100 > 2, 3, 
        ifelse(n_max_50_75 > 2, 2,
        ifelse(n_max_25_50 > 2, 1, 0)))]

        new_row = DataFrame(deepcopy(row))
        new_row.min_impact_class = min_impact_class
        new_row.n_min_75_100 = [n_min_75_100]
        new_row.n_min_50_75 = [n_min_50_75]
        new_row.n_min_25_50 = [n_min_25_50]
        new_row.n_min_0_25 = [n_min_0_25]
        new_row.n_min_0 = [n_min_0]

        new_row.max_impact_class = max_impact_class
        new_row.n_max_75_100 = [n_max_75_100]
        new_row.n_max_50_75 = [n_max_50_75]
        new_row.n_max_25_50 = [n_max_25_50]
        new_row.n_max_0_25 = [n_max_0_25]
        new_row.n_max_0 = [n_max_0]

        append!(impact_classified_roads, new_row)
     end
     
     return impact_classified_roads
end

htf_polygons = SF.st_read("/Users/adam/Downloads/htf_on_rds_int.gpkg")

@time htf_polygons_impact_df = classify_road_flooding(;road_segments=road_segments, htf_polygons=htf_polygons)  

File: htf_impact_analysis.R
library(sf)
library(tidyverse)

road_segments <- sf::st_read("/Users/adam/Downloads/osm_edges.gpkg")

# This will be in a loop
htf_polygons <- sf::st_read("/Users/adam/Downloads/htf_on_rds_int.gpkg")


classify_road_flooding <- function(road_segments, htf_polygons){
  roads_affected <- unique(htf_polygons$id)
  
  filtered_road_segments <- road_segments |> 
    dplyr::select(lanes, highway, id, u, v, geom) |> 
    filter(id %in% roads_affected)
  
  filtered_road_segments$lanes = as.numeric(filtered_road_segments$lanes)
  filtered_road_segments$lanes[is.na(filtered_road_segments$lanes)] <- 2
  filtered_road_segments$buffer_distance <- filtered_road_segments$lanes * 2.5
  
  filtered_road_segments <- filtered_road_segments |> 
    mutate(new_id = paste(id, u, v, sep= "_"))
  
  buffered_filtered_road_segments <- sf::st_buffer(x = filtered_road_segments, dist = filtered_road_segments$buffer_distance)
  
  impacted_buffered_road_segments <- buffered_filtered_road_segments[htf_polygons,]
  
  impacted_road_segments <- filtered_road_segments |> 
    filter(new_id %in% impacted_buffered_road_segments$new_id)
  
  hc_polygons <- htf_polygons |> 
    filter(DN == 1)
  
  lc_polygons <- htf_polygons |> 
    filter(DN == 0)
  p <- progressr::progressor()
  
  roads_with_impact_class <- foreach(i = 1:100, .combine = "bind_rows") %do%{ #nrow(impacted_road_segments)
    p()
    # Select a row and buffered row
    row <- impacted_road_segments[i,]
    buff_row <- impacted_buffered_road_segments[i,]
    
    # calculate slope
    line_end_coords <- sf::st_coordinates(row)
    line_diff <- diff(line_end_coords)[1:2]
    slope <- -1/(line_diff[2] / line_diff[1])
    
    # Break line up into little parts
    segmented_line <- sf::st_segmentize(row, dfMaxLength = 2) |> 
      sf::st_cast("POINT")
    
    # Get the perpendicular lines from the road segment
    perp_lines <- foreach(j = 1:nrow(segmented_line), .combine = "bind_rows") %do% {
      z <- segmented_line[j,]
      
      z_coords <- sf::st_coordinates(z)
      
      x_diff <- z$buffer_distance / sqrt((slope ^ 2) + 1)
      
      sf::st_geometry(z) <- sf::st_sfc(sf::st_linestring(matrix(c(z_coords[1] + x_diff, z_coords[1] - x_diff, z_coords[2] + (x_diff * slope), z_coords[2] - (x_diff * slope)), nrow = 2)), crs = row |> sf::st_crs())
      
      z
    }
    
    # Select high-confidence and low-confidence polygons that intersect the buffered roadway
    hc_polygons_selected <- hc_polygons[buff_row,]|> 
      sf::st_union() 
    
    lc_polygons_selected <- lc_polygons[buff_row,] |> 
      sf::st_union()
    
    # Calculate the length of the line. This should be 2 * buffer_distance
    perp_lines <- perp_lines |> 
      mutate(length = sf::st_length(perp_lines))
    
    # Measure length of perpendicular line with HIGH-confidence htf erased. How much road is not covered
    if(length(hc_polygons_selected) > 0){
      perp_lines <- perp_lines |> 
        mutate(hc_length = sf::st_length(sf::st_difference(perp_lines, hc_polygons_selected)))
    }
    if(length(hc_polygons_selected) == 0){
      perp_lines <- perp_lines |> 
        mutate(hc_length = length)
    }
    
    # Measure length of perpendicular line with LOW-confidence htf erased. How much road is not covered
    if(length(lc_polygons_selected) > 0){
      perp_lines <- perp_lines |> 
        mutate(lc_length = sf::st_length(sf::st_difference(perp_lines, lc_polygons_selected)))
    }
    if(length(lc_polygons_selected) == 0){
      perp_lines <- perp_lines |> 
        mutate(lc_length = length)
    }
    
    perp_lines <- perp_lines |> 
      mutate(hc_percent_covered = 100 - (units::drop_units(hc_length/length) * 100),
             lc_percent_covered = 100 - (units::drop_units(lc_length/length) * 100),
             min_percent_covered = hc_percent_covered,
             max_percent_covered = hc_percent_covered + lc_percent_covered)
    
    
    n_min_75_100 = sum(perp_lines$min_percent_covered >= 75)
    n_min_50_75 = sum(perp_lines$min_percent_covered >= 50 & perp_lines$min_percent_covered < 75)
    n_min_25_50 = sum(perp_lines$min_percent_covered >= 25 & perp_lines$min_percent_covered < 50)
    n_min_0_25 = sum(perp_lines$min_percent_covered > 0 & perp_lines$min_percent_covered < 25)
    n_min_0 = sum(perp_lines$min_percent_covered == 0.0)
    
    n_max_75_100 = sum(perp_lines$max_percent_covered >= 75)
    n_max_50_75 = sum(perp_lines$max_percent_covered >= 50 & perp_lines$max_percent_covered < 75)
    n_max_25_50 = sum(perp_lines$max_percent_covered >= 25 & perp_lines$max_percent_covered < 50)
    n_max_0_25 = sum(perp_lines$max_percent_covered > 0 & perp_lines$max_percent_covered < 25)
    n_max_0 = sum(perp_lines$max_percent_covered == 0.0)
    
    # Impact class: 3 = 75-100% coverage of HTF on road, 2 = 50 - 75%, 1 = 25 - 50%, 0 = 0 - 25%
    min_impact_class = ifelse(n_min_75_100 > 2, 3, 
                               ifelse(n_min_50_75 > 2, 2,
                                      ifelse(n_min_25_50 > 2, 1, 0)))
    
    max_impact_class = ifelse(n_max_75_100 > 2, 3, 
                               ifelse(n_max_50_75 > 2, 2,
                                      ifelse(n_max_25_50 > 2, 1, 0)))
    
    row <- row |> 
      mutate(min_impact_class = min_impact_class,
             n_min_75_100 = n_min_75_100,
             n_min_50_75 = n_min_50_75,
             n_min_25_50 = n_min_25_50,
             n_min_0_25 = n_min_0_25,
             n_min_0 = n_min_0,
             max_impact_class = max_impact_class,
             n_max_75_100 = n_max_75_100,
             n_max_50_75 = n_max_50_75,
             n_max_25_50 = n_max_25_50,
             n_max_0_25 = n_max_0_25,
             n_max_0 = n_max_0)
    
    row
  }

  return(roads_with_impact_class)
  
}
